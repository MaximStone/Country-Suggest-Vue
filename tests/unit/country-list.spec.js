import CountrySuggest from '@/components/Country-Suggest.vue'
import { shallowMount } from '@vue/test-utils'

let testProps = { apiURL: 'https://restcountries.eu/rest/v2/name/{name}?fields=name;flag;',
  flagField: 'flag',
  flagFieldFunction: function (country) {
    return country.flag
  } }

describe('Country Suggestion input tests', () => {
  it('Empty state after render', () => {
    const wrapper = shallowMount(CountrySuggest, {
      propsData: {
        ...testProps
      }
    })
    expect(wrapper.vm.suggestions).toHaveLength(0)
    expect(wrapper.findAll('li.autocomplete-result').length).toBe(0)
  })

  it('Components renders correctly', () => {
    const wrapper = shallowMount(CountrySuggest, {
      propsData: {
        ...testProps
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Props is correct', () => {
    const wrapper = shallowMount(CountrySuggest, {
      propsData: {
        ...testProps
      }
    })
    expect(wrapper.vm._props.apiURL).toBe(testProps.apiURL)
    expect(wrapper.vm._props.flagField).toBe(testProps.flagField)
  })

  it('Suggestions for "German"', (done) => {
    const wrapper = shallowMount(CountrySuggest, {
      propsData: {
        ...testProps
      }
    })
    const textInput = wrapper.find('input[type="text"]')
    textInput.setValue('German')

    setTimeout(() => {
      expect(wrapper.vm.suggestions).toHaveLength(1)
      const options = wrapper.findAll('li.autocomplete-result')
      expect(options.at(0).text()).toBe('Germany')
      expect(wrapper.element).toMatchSnapshot()
      done()
    }, 1000)
  })

  it('Use function type for flagField', (done) => {
    const wrapper = shallowMount(CountrySuggest, {
      propsData: {
        flagField: testProps.flagFieldFunction
      }
    })
    const textInput = wrapper.find('input[type="text"]')
    textInput.setValue('German')

    setTimeout(() => {
      expect(wrapper.vm.suggestions).toHaveLength(1)
      const option = wrapper.find('li.autocomplete-result img')
      expect(option.attributes('src')).toBe('https://restcountries.eu/data/deu.svg')
      done()
    }, 1000)
  })

  it('Suggestions for "sw"', (done) => {
    const wrapper = shallowMount(CountrySuggest, {
      propsData: {
        ...testProps
      }
    })

    const textInput = wrapper.find('input[type="text"]')
    textInput.setValue('sw')

    setTimeout(() => {
      expect(wrapper.vm.suggestions).toHaveLength(4)
      const options = wrapper.findAll('li.autocomplete-result')
      expect(options.at(0).text()).toBe('Botswana')
      expect(options.at(0).contains('img')).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
      done()
    }, 1000)
  })

  it('Default apiURL and flagField works when props is empty', () => {
    const wrapper = shallowMount(CountrySuggest)
    expect(wrapper.vm._props.apiURL).toBe(testProps.apiURL)
    expect(wrapper.vm._props.flagField).toBe(testProps.flagField)
    expect(wrapper.vm.suggestions).toHaveLength(0)
    expect(wrapper.findAll('li.autocomplete-result').length).toBe(0)
  })
})
