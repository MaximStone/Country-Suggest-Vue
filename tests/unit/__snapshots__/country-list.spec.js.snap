// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country Suggestion input tests Components renders correctly 1`] = `
<div
  class="autocomplete"
>
  <input
    type="text"
  />
   
  <ul
    class="autocomplete-results"
    style="display: none;"
  />
</div>
`;

exports[`Country Suggestion input tests Suggestions for "German" 1`] = `
<div
  class="autocomplete"
>
  <input
    type="text"
  />
   
  <ul
    class="autocomplete-results"
    style=""
  >
    <li
      class="autocomplete-result"
    >
      <img
        src="https://restcountries.eu/data/deu.svg"
      />
       
      <span>
        <span
          "=""
          class="bold"
        >
          German
        </span>
        y
      </span>
    </li>
  </ul>
</div>
`;

exports[`Country Suggestion input tests Suggestions for "sw" 1`] = `
<div
  class="autocomplete"
>
  <input
    type="text"
  />
   
  <ul
    class="autocomplete-results"
    style=""
  >
    <li
      class="autocomplete-result"
    >
      <img
        src="https://restcountries.eu/data/bwa.svg"
      />
       
      <span>
        Bot
        <span
          "=""
          class="bold"
        >
          sw
        </span>
        ana
      </span>
    </li>
    <li
      class="autocomplete-result"
    >
      <img
        src="https://restcountries.eu/data/swz.svg"
      />
       
      <span>
        <span
          "=""
          class="bold"
        >
          Sw
        </span>
        aziland
      </span>
    </li>
    <li
      class="autocomplete-result"
    >
      <img
        src="https://restcountries.eu/data/swe.svg"
      />
       
      <span>
        <span
          "=""
          class="bold"
        >
          Sw
        </span>
        eden
      </span>
    </li>
    <li
      class="autocomplete-result"
    >
      <img
        src="https://restcountries.eu/data/che.svg"
      />
       
      <span>
        <span
          "=""
          class="bold"
        >
          Sw
        </span>
        itzerland
      </span>
    </li>
  </ul>
</div>
`;
